File: demo/java-demo/src/main/java/com/alibaba/testable/demo/DemoService.java
package com.alibaba.testable.demo;

import org.springframework.stereotype.Service;
import sun.net.www.http.HttpClient;

import java.net.URL;

@Service
public class DemoService {

    private int count;

    /**
     * Target 1 - private method
     */
    private String privateFunc(String s, int i) {
        return s + " - " + i;
    }

    /**
     * Target 2 - method with private field access
     */
    public String privateFieldAccessFunc() {
        count += 2;
        return String.valueOf(count);
    }

    /**
     * Target 3 - method with new operation
     */
    public String newFunc() {
        BlackBox component = new BlackBox("something");
        return component.get();
    }

    /**
     * Target 4 - method with member method invoke
     */
    public String outerFunc(String s) throws Exception {
        return "{ \"res\": \"" + innerFunc(s) + "\"}";
    }

    /**
     * Target 5 - method with common method invoke
     */
    public String commonFunc() {
        return "anything".trim() + "__" + "anything".substring(1, 2) + "__" + "abc".startsWith("ab");
    }

    /**
     * Target 6 - method with static method invoke
     */
    public BlackBox getBox() {
        return BlackBox.secretBox();
    }

    /**
     * Target 7 - method with override method invoke
     */
    public Box putBox() {
        Box box = new BlackBox("");
        box.put("data");
        return box;
    }

    /**
     * Target 8 - two methods invoke same private method
     */
    public String callerOne() {
        return callFromDifferentMethod();
    }

    public String callerTwo() {
        return callFromDifferentMethod();
    }

    private String innerFunc(String s) throws Exception {
        return HttpClient.New(new URL("http:/xxx/" + s)).getURLFile();
    }

    private String callFromDifferentMethod() {
        return "realOne";
    }

}

File: demo/kotlin-demo/src/main/kotlin/com/alibaba/testable/demo/DemoService.kt
package com.alibaba.testable.demo

import org.springframework.stereotype.Service
import sun.net.www.http.HttpClient
import java.net.URL


@Service
class DemoService {

    private var count = 0

    /**
     * Target 1 - private method
     */
    private fun privateFunc(s: String, i: Int): String {
        return "$s - $i"
    }

    /**
     * Target 2 - method with private field access
     */
    fun privateFieldAccessFunc(): String {
        count += 2
        return count.toString()
    }

    /**
     * Target 3 - method with new operation
     */
    fun newFunc(): String {
        return BlackBox("something").callMe()
    }

    /**
     * Target 4 - method with member method invoke
     */
    fun outerFunc(s: String): String {
        return "{ \"res\": \"" + innerFunc(s) + "\"}"
    }

    /**
     * Target 5 - method with common method invoke
     */
    fun commonFunc(): String {
        val box = BlackBox("anything")
        return box.trim() + "__" + box.substring(1, 2) + "__" + box.startsWith("any")
    }

    /**
     * Target 6 - method with static method invoke
     */
    fun getBox(): BlackBox {
        return ColorBox.createBox("Red", BlackBox.secretBox())
    }

    fun callerOne(): String {
        return callFromDifferentMethod()
    }

    fun callerTwo(): String {
        return callFromDifferentMethod()
    }

    private fun innerFunc(s: String) = HttpClient.New(URL("http:/xxx/$s")).urlFile

    private fun callFromDifferentMethod() = "realOne"
}
